src/persist/src/indexed/encoding.rs

```diff
@@ -100,18 +100,45 @@ pub struct BlobFutureMeta {
 ///
 /// Invariants:
 /// - The batch Descriptions are sorted, non-overlapping, and contiguous.
+/// - The since frontier is either 0 or < the trace's upper (the largest upper
+///   amongst all of the batches).
+/// - Every batch's since frontier is <= the overall trace's since frontier.
+/// - The compaction level of batches is weakly decreasing when iterating from oldest
+///   to most recent time intervals.
+/// - TODO: key uniqueness invariants?
 #[derive(Clone, Debug, Abomonation)]
 pub struct BlobTraceMeta {
     /// The stream this trace belongs to.
     pub id: Id,
-    /// The batches that make up the BlobTrace, represented by their description
-    /// and the key to retrieve the batch's data from the blob store. Note that
-    /// Descriptions are half-open intervals `[lower, upper)`.
-    pub batches: Vec<(Description<u64>, String)>,
+    /// Metadata about he batches that make up the BlobTrace.
+    pub batches: Vec<BlobTraceBatchMeta>,
+    /// Compaction frontier for the batches contained in this trace.
+    /// There may still be batches containing updates at times < since, but the
+    /// the trace only contains correct answers for times at or in advance of this
+    /// of this frontier. Readers are expected to advance any updates < since to
+    /// since.
+    pub since: Antichain<u64>,
     /// The next id used to assign a Blob key for this trace.
     pub next_blob_id: u64,
 }
 
+/// The metadata necessary to reconstruct a [BlobTraceBatch].
+///
+/// Invariants:
+/// - The Description's time interval is non-empty.
+/// - Compaction level is nonzero.
```

Comment:
why don't the levels start counting from 0? if there's some reason for this that i'm not seeing, we should probably spell it out in a comment

Reply:
🤦🏾 yeah ofc that makes more sense